pkg/matrix/matrix.go

```diff
@@ -18,9 +18,9 @@ import (
 )
 
 // FanOut produces combinations of Parameters of type String from a slice of Parameters of type Array.
-func FanOut(params []v1beta1.Param) Combinations {
+func FanOut(matrix v1beta1.Matrix) Combinations {
```

Comment:
Can we please declare this function as a member function:

```
func (m Matrix) FanOut() Combinations {
	var combinations Combinations
	for _, parameter := range m.Params {
		combinations = combinations.fanOut(parameter)
	}
	return combinations
}
```

just like:

```
func (combinations Combinations) fanOut(param v1beta1.Param) Combinations {
```

Reply:
/hold for now until we clarify how this works since the existing logic has matrix.FanOut()

matrixCombinations := matrix.FanOut(rpt.PipelineTask.Matrix.Params).ToMap()

Reply:
Chatted with Priti and decided for now we will leave as is. The way that Matrix pkg is right now does not make it is easy to simply refactor `FanOut()` as a member function.
/hold cancel
/assign @pritidesai

Reply:
Yes. We would have to move `FanOut()` to where the type Matrix is defined in `Pipeline_types.go`, however Combinations is defined within Matrix and it doesn't make sense to have Combinations defined in the API.

Reply:
thanks @EmmaMunley!

Yup, it results in cyclic imports 🙃 and bringing `combinations` in `pkg/apis/` might not be desirable based on recent discussions of avoiding copy/paste in multiple APIs.

@jerop I am happy to merge this for now but wanted to understand this package `github.com/tektoncd/pipeline/pkg/matrix`. Also, the package name `matrix` being same as the type `Matrix` caused confusion in https://github.com/tektoncd/pipeline/blob/main/pkg/reconciler/pipelinerun/pipelinerun.go#L837

```
matrix.FanOut(rpt.PipelineTask.Matrix.Params).ToMap()
```

Reply:
Matrix type is defined in `pkg/apis/pipeline/v1beta1/` with a couple of member functions. Matrix type has two fields: Params []Param and Include []MatrixInclude

https://github.com/tektoncd/pipeline/blob/main/pkg/apis/pipeline/v1beta1/pipeline_types.go#L165

https://github.com/tektoncd/pipeline/blob/90f33a63497e8c07fa799f8cab5138756a075299/pkg/apis/pipeline/v1beta1/pipeline_types.go#L329

https://github.com/tektoncd/pipeline/blob/90f33a63497e8c07fa799f8cab5138756a075299/pkg/apis/pipeline/v1beta1/pipeline_types.go#L334

There is a matrix package with `matrix.go` and `matrix_types.go` with a `combinations` defined. Combination type has two fields: MatrixID string and Params []v1beta1.Param.

Can we replace `combination` type with matrix struct and add MatrixID to matrix struct? This way we do not have maintain a separate struct and a package.

Reply:
@pritidesai thanks for the suggestions, refactored the implementation - https://github.com/tektoncd/pipeline/pull/6282